commonmark-ext-metadata/src/main/java/org/commonmark/ext/metadata/internal/MetadataBlockParser.java

```diff
@@ -0,0 +1,147 @@
+                    node.setKey(key);
+                    node.setValues(values);
+                    block.appendChild(node);
+                }
+
+                literal = false;
+                key = matcher.group(1);
+                values = new ArrayList<>();
+                if ("|".equals(matcher.group(2))) {
+                    literal = true;
+                } else if (!"".equals(matcher.group(2))) {
+                    values.add(matcher.group(2));
+                }
+            } else {
+                if (literal) {
+                    matcher = REGEX_METADATA_LITERAL.matcher(line);
+                    if (matcher.matches()) {
+                        if (values.size() == 1) {
+                            values.set(0, values.get(0) + "\n" + matcher.group(1).trim());
+                        } else {
+                            values.add(matcher.group(1).trim());
+                        }
+                    }
+                } else {
+                    matcher = REGEX_METADATA_LIST.matcher(line);
+                    if (matcher.matches()) {
+                        values.add(matcher.group(1));
+                    }
+                }
+            }
+        }
+
+        if (key != null) {
+            MetadataNode node = new MetadataNode();
+            node.setKey(key);
+            node.setValues(values);
+            block.appendChild(node);
+        }
+    }
+
+    public static class Factory extends AbstractBlockParserFactory {
+        @Override
+        public BlockStart tryStart(ParserState state, MatchedBlockParser matchedBlockParser) {
+            CharSequence line = state.getLine();
+            BlockParser parentParser = matchedBlockParser.getMatchedBlockParser();
+            // check whether this line is the first line of whole document or not
+            if (parentParser instanceof DocumentBlockParser && parentParser.getBlock().getFirstChild() == null &&
+                    REGEX_BOUNDARY.matcher(line).matches()) {
+                // count valid metadata line in metadata block
+                int prevIndex;
+                int index = nextLineEnd(line, 0);
+                int validLineCount = 0;
+                CharSequence subseq;
+                do {
+                    prevIndex = index + 1;
+                    index = nextLineEnd(line, prevIndex);
+                    subseq = line.subSequence(prevIndex, index);
+                    if (REGEX_METADATA.matcher(subseq).matches()) {
+                        validLineCount++;
+                    }
+                } while (!REGEX_BOUNDARY.matcher(subseq).matches() && prevIndex != index);
+
+                if (validLineCount > 0) {
+                    return BlockStart.of(new MetadataBlockParser()).atIndex(state.getNextNonSpaceIndex());
+                }
+            }
+
+            return BlockStart.none();
+        }
+
+        private int nextLineEnd(CharSequence seq, int startIndex) {
```

Comment:
If we check only one line to decide whether metadata block starts or not, there are some collisions with CommonMark spec like following case:

```
---
test
----
```

From original CommonMark spec, rendering result should be `<hr />\n<h2>test</h2>\n`. But because this extension consider `---` as start of metadata block, the actual result will be empty. Is there any method to use multi-line input for detecting start of the block?